vcscommit.man
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH VCSCOMMIT "1" "November 2009" "vcscommit $VER$" "manual page for vcsommit"
.SH NAME
vcscommit - Comment on a bug if the message contains a Fixes: NB#<BUGN> string
.SH DESCRIPTION
Parses the changelog contained in a commit message (either SVN, GIT or HG) and then comments on bugzilla the relevant bug
.SS "options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-t\fR TRAC, \fB\-\-trac\fR=\fITRAC\fR
Path to the Trac project
.TP
\fB\-G\fR GIT, \fB\-\-git\fR=\fIGIT\fR
Git url (used for dvcs) conflicts with \fB\-U\fR and with \fB\-H\fR
.TP
\fB\-H\fR HG, \fB\-\-hg\fR=\fIHG\fR
HG url (used for dvcs) conflicts with \fB\-U\fR and with \fB\-G\fR
.TP
\fB\-U\fR URL, \fB\-\-url\fR=\fIURL\fR
URL to use in changeset references
.TP
\fB\-r\fR REV, \fB\-\-revision\fR=\fIREV\fR
Repository revision number
.TP
\fB\-T\fR TAG, \fB\-\-tag\fR=\fITAG\fR
Tag information
.TP
\fB\-u\fR USER, \fB\-\-user\fR=\fIUSER\fR
The user who is responsible for this action
.TP
\fB\-m\fR MSG, \fB\-\-msg\fR=\fIMSG\fR
The log message to search
.TP
\fB\-c\fR ENCODING, \fB\-\-encoding\fR=\fIENCODING\fR
The encoding used by the log message
.TP
\fB\-b\fR BUGZILLA, \fB\-\-bugzilla\fR=\fIBUGZILLA\fR
Bugzilla base URL (mandatory)
.TP
\fB\-\-rest_uri\fR=\fIREST_URI\fR
Provides the REST part of the URL for the bugzilla server (mandatory)
.TP
\fB\-P\fR PROXY, \fB\-\-proxy\fR=\fIPROXY\fR
Proxy server address
.TP
\fB\-n\fR NETRC, \fB\-\-netrc\fR=\fINETRC\fR
Netrc file name
.TP
\fB\-v\fR, \fB\-\-version\fR
show the version and exit
.SH EXAMPLES
\fBWith SVN\fR
.TP
You need to include something like this into your post-commit hook:

 REPOS="$1"
 REV="$2"
 LOG=`/usr/bin/svnlook log -r $REV $REPOS`
 AUTHOR=`/usr/bin/svnlook author -r $REV $REPOS`

 /usr/bin/python /path/to/pmo-commit \\
  -U "<URL/TO/TRAC>"  \\
  -r "$REV"       \\
  -u "$AUTHOR"    \\
  -m "$LOG"
.TP
\fBWith GIT || HG\fR
.TP
You should use python-vcs-post-receive that ships a post-receive hook ad-hoc.

.SH AUTHORS
Mikhail Sobolev <mikhail.sobolev@nokia.com>, Stefano Mosconi <stefano.mosconi@gmail.com>
.SH MAINTAINER
Stefano Mosconi <stefano.mosconi@gmail.com>
.SH COPYRIGHT
Copyright (C) 2008 Nokia Corporation. All rights reserved.

The contents of this file are subject to the Mozilla Public License
Version 1.1 (the "License"); you may not use this file except in
compliance with the License. You may obtain a copy of the License at
http://www.mozilla.org/MPL/

Software distributed under the License is distributed on an "AS IS"
basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
License for the specific language governing rights and limitations
under the License.
.SH "SEE ALSO"
python-vcs-post-receive(1)


debian/changelog
python-vcs-commit (0.6.2) unstable; urgency=low

  * Bugfixing
  * Correct python-minideblib dependency

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Wed, 28 Sep 2011 13:46:01 +0300

python-vcs-commit (0.6.1) unstable; urgency=low

  * Forcing old pkg removal 

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Wed, 14 Sep 2011 15:09:59 +0300

python-vcs-commit (0.6) unstable; urgency=low

  * Using bugzilla REST api now
  * Renaming package to python-vcs-commit to have more general meaning and 
    compliance with python standards
  * updating packaging
  * giving it an object oriented structure
  * slightly improving options handling
  * start to have a config file
  * Fixes: NB#281560 - Automatic commit comments environment is expanded

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Mon, 12 Sep 2011 22:47:30 +0300

python-vcs-commit (0.5.2) unstable; urgency=low

  * Version bump due to package building mess.

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Tue, 12 Jul 2011 17:03:10 +0300

python-vcs-commit (0.5.1) unstable; urgency=low

  * Fixes: NB#201475 -  Writing only the part of the changelog
    that is relevant for a given bug

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Sun, 07 Nov 2010 13:59:27 +0200

python-vcs-commit (0.5) jaunty; urgency=low

  *  Re-engineering the whole thing
  *  Tag is not used anymore
  *  Support for changelog introduced
  *  Changelog is preferred to commit message and gets package name and
     version

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Mon, 13 Sep 2010 14:10:00 +0300

python-vcs-commit (0.4.7) jaunty; urgency=low

  *  import os was missing

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Wed, 28 Jul 2010 10:03:31 +0300

python-vcs-commit (0.4.6) unstable; urgency=low

  * Removing revision, adding tag information

 -- Saranszky Csaba <ext-csaba.saranszky@nokia.com>  Tue, 22 Jun 2010 11:36:30 +0300

python-vcs-commit (0.4.5) unstable; urgency=low

  * Adding HG capability

 -- Saranszky Csaba <ext-csaba.saranszky@nokia.com>  Wed, 16 Jun 2010 11:16:14 +0300

python-vcs-commit (0.4.4) unstable; urgency=low

  * Supporting also multiple bugfixes in the same changelog now

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Wed, 02 Jun 2010 17:15:35 +0300

python-vcs-commit (0.4.3.1) unstable; urgency=low

  * Minor linguisting change

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Thu, 29 Apr 2010 17:18:44 +0300

python-vcs-commit (0.4.3) jaunty; urgency=low

  * Adding WAITING and NEED_INFO to the possible bug_statuses to be resolved

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Tue, 27 Apr 2010 15:55:33 +0300

python-vcs-commit (0.4.2) unstable; urgency=low

  * Adapting code to new bugzilla 3.4.x

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Fri, 15 Jan 2010 11:53:08 +0200

python-vcs-commit (0.4.1) unstable; urgency=low

  * Bugfix 

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Fri, 11 Dec 2009 17:02:59 +0200

python-vcs-commit (0.4) unstable; urgency=low

  * Adding option to comment only 

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Fri, 11 Dec 2009 16:39:17 +0200

python-vcs-commit (0.3) unstable; urgency=low

  * Adding option to set the TM of a bug when closing it 

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Fri, 06 Nov 2009 12:50:30 +0200

python-vcs-commit (0.2) unstable; urgency=low

  * Proper dependencies in place

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Wed, 04 Nov 2009 13:45:25 +0200

python-vcs-commit (0.1) unstable; urgency=low

  * First debianization

 -- Stefano Mosconi <stefano.mosconi@gmail.com>  Wed, 04 Nov 2009 09:38:31 +0200


vcscommit.py
#! /usr/bin/python

import sys, netrc, urllib, os

from optparse import OptionParser
from datetime import datetime
from ConfigParser import ConfigParser

from minideblib.DpkgChangelog import DpkgChangelogEntry, DpkgChangelog
from bugzillarest import BugzillaREST

class VCStoBugzilla():
    """
    An easy way to process debian changelog messages from git/hg/svn and commit them to Bugzilla

    Either comment only or resolve/fix a bug if specific keywords are present
    """

    def __init__(self, options):
        """
        Initialization and check of the options
        
        :param options: an object that holds all the info needed for the hook to proceed
        :type options: class

        options needs to have all these properties

        :param bugzilla: The bugzilla server URL (without the rest uri)
        :type bugzilla: string
        :param rest_uri: The REST part of the url
        :type rest_uri: string
        :param chglog: The changelog message (extracted from debian/changelog)
        :type chglog: string
        :param commentonly: Do only comment
        :type commentonly: bool
        :param msg: The commit message
        :type msg: string
        :param netrc: The netrc file location
        :type netrc: string
        :param proxy: Overrides the proxy in case you don't like the ENV one
        :type proxy: string
        :param rev: The revision number / commit hash
        :type rev: string
        :param tm: Whether to add the target milestone to the bug (as the current week)
        :type tm: bool
        :param user: The user that did the commit (coming from the VCS)
        :type user: string
        :param vcstype: What VCS are we working with (HG/GIT/SVN)
        :type vcstype: string
        :param vcsurl: the base url of the VCS (this is not trivial to guess, look at the code)
        :type vcsurl: string
        """
        supported_vcstypes = ['hg', 'git', 'trac', 'svn']
        if options.vcstype not in supported_vcstypes:
            print >> sys.stderr, "Unsupported vcs type. Supported types are %s " % supported_vcstypes
        self.options = options
        self.config = self.parse_config()
        if self.options.vcstype == 'svn':
            self.finalurl = os.path.join(self.options.vcsurl, self.config['svn_commiturl'], self.options.rev)
        elif self.options.vcstype == 'git':
            self.finalurl = self.options.vcsurl + self.config['git_commit_url'] % self.options.rev
        elif self.options.vcstype == 'hg':
            self.finalurl = os.path.join(self.options.vcsurl, self.config['hg_commit_url'], '%s' % self.options.rev)
        elif self.options.vcstype == 'trac':
            from trac.env import open_environment
            env = open_environment(self.options.vcsurl)
            self.finalurl = os.path.join(env.config.get('project', 'url'), self.config['trac_commit_url'], self.options.rev)
        else:
            print >> sys.stderr, 'Configuration is not complete: please check the options passed'
            sys.exit(1)

        if self.options.tm:
            year, week = datetime.now().isocalendar()[0:2]
            self.target_milestone = '%d-%02d' % (year, week)
        else:
            self.target_milestone = None
        self.open_statuses = [ 'NEW', 'ASSIGNED', 'REOPENED', 'WAITING', 'NEED_INFO' ]

    def parse_config(self):
        """
        Reads the config file (default in /etc/vcscommit/vcscommit.cfg

        :rtype: ConfigParser object
        """
        cfg = ConfigParser()
        cfg.read('/etc/vcscommit/vcscommit.cfg')
        return cfg.defaults()

    def validate_login(self, url, netrcfile=None):
        """
        Takes username and pwd from netrc file

        :param url: the url we are authenticating to
        :type url: string
        :param netrcfile: the location of the netrc file
        :type netrcfile: string
        """
        base_host = urllib.splithost(urllib.splittype(url)[1])[0]
        # Try to grab authenticators out of your .netrc
        try:
            name, account, password = netrc.netrc(netrcfile).authenticators(base_host)
        except:
            raise SystemExit("You must have valid ~/.netrc for host %s" % base_host)

        return name, account, password

    def work_the_bug(self, bugs, message):
        """
        Does the actual comment on a set of bugs with the given message

        :param bugs: List of bugs to work on
        :type bugs: list
        :param message: Message to write on the bug
        :type message: string
        """
        for bug in bugs:
            retries = 0
            res = False
            while not res and retries <= 3:
                if retries > 0:
                    print sys.stdout, "Something went wrong with Bugzilla retrying (%s)" % retries
                if not self.options.commentonly:
                    res = bug.resolve(message, 'fixed', self.target_milestone)
                else:
                    res = bug.comment(message)
                retries += 1

            if not res:
                print >> sys.stdout, "Sorry we failed to update bug %s" % bug.id
                print >> sys.stderr, res
            else:
                print >> sys.stdout, "Bug #%s processed" % bug.id

    def process_entry(self, entry):
        """
        Process one entry in the changelog
        
        :param entry: The changelog entry
        :type entry: DpkgChangelogEntry object
        """

        if entry.nbugsfixed:
            name, account, password = self.validate_login(self.options.bugzilla, self.options.netrc)

            bugzilla = {'base_url': self.options.bugzilla,
                        'rest_uri': self.options.rest_uri,
                        'user': name,
                        'passwd': password}
            if self.options.proxy:
                session = BugzillaREST(bugzilla, self.options.proxy)
            else:
                session = BugzillaREST(bugzilla)

            bugs = session.get_many(entry.nbugsfixed)
            
            if not self.options.commentonly:
                action = 'Fixed'
            else:
                action = 'Commented'

            if entry.package:
                package = "package %s" % entry.package
            else:
                package = ""

            if entry.version:
                version = package + " (%s)" % entry.version.__str__()
            else:
                version = package

            msg_tmpl = '%(action)s as %(user)s submitted %(ver)s:\n'\
                       '%(url)s\n'\
                       'Commit message:\n'\
                       '%(msg)s\n'\
                       '%(chglog)s'
            # Write a different changelog per each bug
            # with relevant information only
            if bugs:
                if self.options.chglog:
                    for bug in bugs:
                        if bug.info.status not in self.open_statuses:
                            print "Bug #%s is in %s status... doing nothing" % (bug.id, bug.info.status)
                            continue
                        for chglog_entry in entry.entries:
                            if chglog_entry.find(str(bug.id)) != -1:
                                the_changelog = "Relevant changelog message:\n%s" % chglog_entry 
                                message = msg_tmpl % {
                                        'ver': version,
                                        'action': action,
                                        'user': self.options.user,
                                        'url' : self.finalurl,
                                        'msg' : self.options.msg,
                                        'chglog' : the_changelog    
                                        }
                                self.work_the_bug([bug], message)    
                                break
                else:
                    the_changelog = ''
                    message = msg_tmpl % {
                        'ver': version,
                        'action': action,
                        'user': self.options.user,
                        'url' : self.finalurl,
                        'msg' : self.options.msg,
                        'chglog' : the_changelog    
                        }
                    self.work_the_bug(bugs, message)    
            
    def run(self):
        """
        Runs the business
        """
        entry = DpkgChangelogEntry()
        for aline in self.options.msg.split('\n'):
            entry.add_entry(aline)


        if self.options.chglog:
            # If we have the changelog we prefer 
            # to use this instead of the commit msg
            ch = DpkgChangelog()
            ch.parse_changelog(self.options.chglog)
            for oneentry in ch.entries:
                self.process_entry(oneentry)
        else:
            self.process_entry(entry)
        

def main():
    parser = OptionParser()
    parser.add_option('-t', '--trac', dest='trac', help='Path to the Trac project')
    parser.add_option('-G', '--git', help='Git url (used for dvcs) conflicts with -U and -H')
    parser.add_option('-H', '--hg', help='HG url (used for dvcs) conflicts with -U and -G')#
    parser.add_option('-U', '--url', dest='url', help='URL to use in changeset references')
    parser.add_option('-r', '--revision', dest='rev', help='Repository revision number')
    parser.add_option('-u', '--user', dest='user', help='The user who is responsible for this action')
    parser.add_option('-m', '--msg', dest='msg', help='The log message to search')
    parser.add_option('-c', '--encoding', dest='encoding', help='The encoding used by the log message')
    parser.add_option('-b', '--bugzilla', dest='bugzilla', help='Bugzilla base URL')
    parser.add_option('--rest', dest='rest_uri', help='Bugzilla REST uri (the .._rest part of the URL')
    parser.add_option('-P', '--proxy', dest='proxy', help='Proxy server address')
    parser.add_option('-n', '--netrc', dest='netrc', default=None, help='Netrc file name')
    parser.add_option('--chglog', dest='chglog', default=None, help='here goes the whole changelog from debian/changelog')
    parser.add_option('--tm', dest='tm', action="store_true", default=False, help='Set target milestone when closing')
    parser.add_option('--commentonly', action="store_true", default=False, help='Comment only, don\'t resolve the bug')

    options, args = parser.parse_args()

    if (options.url and options.git) or (options.git and options.trac) or (options.url and options.trac) or (options.url and options.hg) or (options.hg and options.trac) or (options.hg and options.git):
        print >> sys.stderr, 'Conflict in -U and -G options or -G and -t or -t and -U or -U and -H or -H and it or -H and -G'
        sys.exit(1)


    if not options.bugzilla or not options.rest_uri:
        print >> sys.stderr, "You need to provide a bugzilla url with a --rest uri"
        sys.exit(1)

    # from trac.util.text import to_unicode
    #msg = to_unicode(options.msg, options.encoding)

    VCStoBugzilla(options).run()
    
    if options.url:
        options.vcstype = 'svn'
        options.vcsurl = options.url
    elif options.git:
        options.vcstype = 'git'
        options.vcsurl = options.git
    elif options.hg:
        options.vcstype = 'hg'
        options.vcsurl = options.hg
    elif options.trac:
        options.vcstype = 'trac'
        options.vcsurl = options.trac



if __name__ == '__main__':
    main()

# vim:ts=4:sw=4:et
